README.md
# dsrTest [![Build Status](https://travis-ci.org/mnel/dsrTest.svg)](https://travis-ci.org/mnel/dsrTest) [![License](http://img.shields.io/badge/license-GPL%20%28%3E=%202%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-2.0.html) [![CRAN](http://www.r-pkg.org/badges/version/dsrTest)](https://cran.r-project.org/package=dsrTest) [![Downloads](http://cranlogs.r-pkg.org/badges/dsrTest?color=brightgreen)](http://www.r-pkg.org/pkg/dsrTest) [![Coverage Status](https://img.shields.io/codecov/c/github/mnel/dsrTest/master.svg)](https://codecov.io/github/mnel/dsrTest?branch=master)


Tests and Confidence Intervals on Directly Standardized Rates for Several Methods

## Installation

You can install dsrTest from github with:

```R
remotes::install_github("mnel/dsrTest")
```
or from CRAN

```R
install.packages("dsrTest")
```

# Author 

Michael Nelson

# Licence

GPL (>=2)


NEWS.md
# dsrTest 0.1

- Initial Version

# Version 0.2

- Bug Fixes on alternative = "less" or "greater"
- Improved documentation formatting
- Example data and vignette added
- wmtype modifications added for Beta methods

# Version 0.2.1

- modifactions stop workaround now asht bug fixes are implemented

# Version 0.2.3

- use of testthat now "tested" (suggests != depends)

tests/testthat.R
if(require(testthat, quietly = TRUE)){
  library(dsrTest)
  test_check("dsrTest")
}


tests/testthat/test-alternatives.R
context("Sense checking confidence intervals")

xfive <- c(0, 8, 63, 112, 262, 295)
nfive <- c(327, 30666, 123419, 149919, 104088, 34392)
ntotal <- c(319933, 931318, 786511, 488235, 237863, 61313)

methods_list  <- list(
  gamma = list(
    list(wmtype = "max"),
    list(midp = TRUE),
    list(wmtype = "tcz"),
    list(wmtype = "mean"),
    list(wmtype = "minmaxavg")),
  asymptotic = list(
    list(trans = "none"),
    list(trans = "log"),
    list(trans = "loglog"),
    list(trans = "logit")),
  dobson = list(
    list(midp = FALSE),
    list(midp = TRUE)),
  beta = list(
    list(wmtype = "none"),
    list(wmtype = "tcz"),
    list(wmtype = "mean"),
    list(wmtype = "minmaxavg"),
    list(wmtype = "max")),
  bootstrap = list(list())
)

all_ts <- mapply(dsrTest::dsrTest,
  method = rep(names(methods_list), times = lengths(methods_list)),
  control = do.call(c, unname(methods_list)),
  MoreArgs = list(mult = 1e5, x = xfive, n = nfive, w = ntotal),
  SIMPLIFY = FALSE)

all_less <- mapply(dsrTest::dsrTest,
  method = rep(names(methods_list), times = lengths(methods_list)),
  control = do.call(c, unname(methods_list)),
  MoreArgs = list(mult = 1e5, x = xfive, n = nfive, w = ntotal,
  alternative = "less"), SIMPLIFY = FALSE)

all_greater <- mapply(dsrTest::dsrTest,
  method = rep(names(methods_list), times = lengths(methods_list)),
  control = do.call(c, unname(methods_list)),
  MoreArgs = list(mult = 1e5, x = xfive, n = nfive, w = ntotal,
  alternative = "greater"), SIMPLIFY = FALSE)

test_that("Estimate is within confidence intervals for all methods", {
  expect_true(all(
    sapply(all_less,
      function(x) findInterval(x$estimate, x$conf.int))) == 1L)
  expect_true(all(
    sapply(all_ts,
      function(x) findInterval(x$estimate, x$conf.int))) == 1L)
  expect_true(all(
    sapply(all_greater,
      function(x) findInterval(x$estimate, x$conf.int))) == 1L)
})

less_value <- mapply(dsrTest::dsrTest,
  method = rep(names(methods_list), times = lengths(methods_list)),
  control = do.call(c, unname(methods_list)),
  null.value = sapply(all_less, function(x) x$conf.int[2] - 0.5),
  MoreArgs = list(mult = 1e5, x = xfive, n = nfive, w = ntotal,
  alternative = "less"), SIMPLIFY = FALSE)

test_that("P value is sensible", {
  expect_true(all(sapply(less_value, function(x) x$p.value > 0.05)))
})

test_that("Confidence Intervals are ordered appropriately", {
  all(!sapply(all_less, function(x) is.unsorted(x$confint)))
  all(!sapply(all_greater, function(x) is.unsorted(x$confint)))
  all(!sapply(all_ts, function(x) is.unsorted(x$confint)))
  })

context("Integer overflow")

xL <- c(0L, 8L, 63L, 112L, 262L, 295L)
nL <- c(327L, 30666L, 123419L, 149919L, 104088L, 34392L)
ntL <- c(319933L, 931318L, 786511L, 488235L, 237863L, 61313L)

test_that("Integer valued data doesn't cause issues", {
  expect_equal(
    all_ts[[1]][-9L], dsrTest(xL, nL, ntL, mult = 1e5)[-9L])
})


R/data-downs.mi.R
#' @title Downs' syndrome cases and of total live births 
#' by maternal age and birth order, Michigan, 1950-1964.
#' 
#' @description
#' This data reproduces table 14.4 in Fleiss (1981) and which is
#' drawn from a large-scale study in Michigan 1950-1964 of the effect
#' of Maternal age and Birth order on Down Syndrome and Leukemia 
#' (Stark and Mantell, 1966).
#' @format
#' This `data.frame` contains the following columns:
#' \describe{
#' \item{\code{Age}}{The age category of the mothers.}
#' \item{\code{BirthOrder}}{The birth order.}
#' \item{\code{Cases}}{The number of cases of Down's Syndrome.}
#' \item{\code{Births}}{The number of live births.}
#' \item{\code{Standard}}{A "standard" population, derived as the total
#' number of births in each age category}
#' }
#' @source 
#' The data were obtained from table 14.4 (p 249) in
#' 
#' Fleiss, JL (1981) *Statistical Methods for Rates and Proportions*, 
#' Wiley, New York.
#' 
#' The original study is
#' 
#' Stark CR and Mantel N (1966) 'Effects of maternal age and 
#' birth order on the risk of mongolism and leukemia' 
#' *J Natl Cancer Inst* **37** (5) 687--698. \doi{doi:10.1093/jnci/37.5.687}
"downs.mi"


R/controlFunctions.R
#' @title Control Function for Dobson Method Confidence Intervals
#' 
#' @description Provides a list of arguments to pass to 
#' \code{\link[exactci]{poisson.exact}}.
#' @seealso \code{\link[exactci]{poisson.exact}}
#' @return a list with values 
#' \item{midp}{}
#' \item{tsmethod}{}
#' @param midp logical, use mid-p values? Currently only permitted
#' where `tsmethod = "central"`.
#' @param tsmethod `character` giving two-sided method
#' @param ... Currently ignored..
#' @export
dobsonControl <- function(midp = FALSE,
                          tsmethod = c("central", "minlike", "blaker"),
                          ...){
  tsmethod <- match.arg(tsmethod)
  list(midp = midp, tsmethod = tsmethod)
}

#' @title Control Function for Asymptotic Method Confidence Intervals
#' 
#' @description Specify the transformation to apply to the distribution
#' of the MLE.
#'
#' @param trans Transformation apply to the MLE distribution. 
#' @param ... Currently ignored.
#' @return
#' A list with values
#' \item{trans}{}
#' @export
asymptoticControl <- function(trans = c("none", "log", "loglog", "logit"), ...){
  trans <- match.arg(trans)
  list(trans = trans)
}

#' @title Control Function for Gamma Method Confidence Intervals
#' @description Provides a list of arguments to pass to 
#' \code{\link[asht]{wspoissonTest}}.
#' @seealso \code{\link[asht]{wspoissonTest}}
#' @param midp logical. Use mid-p confidence distribution method? Currently
#' only implemented where `wmtype = "max"`
#' @param nmc Calculation method when `midp = TRUE`.
#' @param wmtype type of modification for the Gamma confidence interval.
#' @param unirootTolFactor tolerance factor for uniroot where `midp = TRUE`
#' and `nmc = 0`.
#' @param ... Currently ignored.
#' @return
#' A list of arguments to pass to \code{\link[asht]{wspoissonTest}}.
#' 
#' If `midp = TRUE`, with values
#' \item{midp}{}
#' \item{nmc}{}
#' \item{unirootTolFactor}{}
#' 
#' If `midp = FALSE`, with values:
#' \item{\code{wmtype}}{}
#' @export
gammaControl <- function(midp = FALSE, nmc = 0,
                         wmtype = c("max", "mean", "minmaxavg", "tcz"),
                         unirootTolFactor = 1e-06, ...){
  wmtype <- match.arg(wmtype)
  if (midp) {
    RVAL <- list(midp = midp, nmc = nmc,
                 unirootTolFactor = unirootTolFactor)
  } else {
      RVAL <- list(wmtype = wmtype, midp = FALSE)
    }
  RVAL
}

#' Control Function for Beta Method for Confidence Intervals
#'
#' @description Modification to the Beta method. The options are `"none"`
#' or the same modifications as applied to the Gamma Method 
#' (see \code{\link{gammaControl}}) are implemented. `wmtype="none"` and
#' `wmtype="tcz"` have been investigated by Tiwari et al (2006) and
#' Ng et al (2008).
#' @references 
#' Tiwari RC, Clegg LX, & Zou Z (2006). 'Efficient interval estimation 
#' for age-adjusted cancer rates.' 
#' *Statistical Methods in Medical Research* **15**: 547--569. 
#' \doi{doi:10.1177/0962280206070621}
#' 
#' Ng HKT, Filardo, G & Zheng G (2008). 'Confidence interval estimating 
#' procedures for standardized incidence rates.' 
#' *Computational Statistics and Data Analysis* **52** 3501--3516.
#' \doi{doi:10.1016/j.csda.2007.11.004}
#' @param wmtype `character` type of modification to the Beta Confidence
#' Interval
#' @param ... Currently ignored.
#'
#' @return a list with values
#' \item{wmtype}{modification to Beta Confidence Interval to implement}
#'
#' @export
betaControl <- function(wmtype = c("none", "tcz", "max", "mean",
                                   "minmaxavg"), ...) {
  wmtype <- match.arg(wmtype)
  list(wmtype = wmtype)
}


data-raw/example.small.R
#' Fleiss table 14.4 (p 249)
#' 
#' 
#' From Stark CR and Mantel N (1966) 'Effects of maternal age and 
#' birth order on the risk of mongolism and leukemia' 
#' *J Natl Cancer Inst* **37** (5) 687--698. \doi{doi:10.1093/jnci/37.5.687}


downs.mi <- structure(
  list(
    Age = structure(c(1L, 1L, 1L, 1L, 1L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 3L, 
     3L, 3L, 4L, 4L, 4L, 4L, 4L, 5L, 5L, 5L, 5L, 5L, 6L, 7L, 7L, 7L, 
     7L), .Label = c("<20", "20-24", "25-29", "30-34", "35-39", "40-50", 
     "40+"), class = "factor"), 
    BirthOrder = c(1L, 2L, 3L, 4L, 5L, 1L, 2L, 3L, 4L, 5L, 1L, 2L, 3L, 
      4L, 5L, 1L, 2L, 3L, 4L, 5L, 1L, 2L, 3L, 4L, 5L, 1L, 2L, 3L, 4L, 5L), 
    Cases = c(107, 25, 3, 1, 0, 141, 150, 71, 26, 8, 60, 
      110, 114, 64,  63, 40, 84, 103, 89, 112, 39, 82, 108, 
      137, 262, 25,39, 75, 96, 295), 
    Births = c(230061, 72202, 15050, 2293,  327, 329449, 326701, 
      175702, 68800, 30666, 114920, 208667, 207081, 132424, 
      123419, 39487, 83228, 117300, 98301, 149919, 14208, 
      28466, 45026, 46075, 104088, 3052, 5375, 8660, 9834, 34392), 
    Standard = c(319933, 319933, 319933, 319933, 
  319933, 931318, 931318, 931318, 931318, 931318, 786511, 
  786511, 786511, 786511, 786511, 488235, 488235, 488235, 
  488235, 488235, 237863, 237863, 237863, 237863, 237863, 
  61313, 61313, 61313, 61313, 61313)), 
  .Names = c("Age", "BirthOrder",  "Cases", "Births", "Standard"), 
  class = "data.frame", row.names = c(NA, -30L))



R/dsrTest.R
#' @title Tests and Confidence Intervals on Directly Standardized Rates
#' @seealso \code{\link[asht]{wspoissonTest}}, 
#' \code{\link[exactci]{poisson.exact}}, 
#' \code{\link{gammaControl}}, 
#' \code{\link{dobsonControl}}, 
#' \code{\link{asymptoticControl}}, 
#' \code{\link{betaControl}}
#' @description A number of methods have been proposed for calculating
#' confidence intervals for directly standardized rates. Ng et al (2008),
#' compare a number of methods, some of which are implemented here. 
#' The default uses the Gamma method by Fay and Feuer (1997) and 
#' implemented in \code{\link[asht]{wspoissonTest}}. 
#' 
#' @details
#' Five classes of method have been implemented here:
#' 
#' \describe{
#' \item{\code{"gamma"}}{Calls \code{\link[asht]{wspoissonTest}}. By default
#' uses the Gamma Method proposed by Fay and Feuer (1997). 
#' Modifications proposed by Tiwari et al (2006) and Fay and Kim (2017) 
#' also implemented - see \code{\link{gammaControl}}.} 
#' \item{\code{"asymptotic"}}{Using the normal approximation of the 
#' MLE or transformed MLE distribution - see \code{\link{asymptoticControl}}}
#' \item{\code{"dobson"}}{Uses the method proposed by Dobson et al (1991). 
#' Estimating the confidence interval on the unweighted sum is done by calling
#' \code{\link[exactci]{poisson.exact}} - both the exact method and
#' a mid-p method are possible - see \link{dobsonControl}.}
#' \item{\code{"beta"}}{Methods based on the beta distribution by Tiwari et
#' al (2006) - see \link{betaControl}.}
#' \item{\code{"bootstrap"}}{Approximate Bootstrap method by Swift (1995). 
#' P-values are estimated by solving for p.}
#' }
#' 
#' For each method there is a `control` function that will return a list of
#' parameters that can be used to define sub-types of each of the broad groups
#' @references 
#' Dobson, AJ, Kuulasmaa, K, Eberle, E and Scherer, J (1991) 
#' 'Confidence intervals for weighted sums of Poisson parameters', 
#' *Statistics in Medicine*, **10**: 457--462.
#' \doi{doi:10.1002/sim.4780100317}
#' 
#' Swift, MB (1995). 'Simple confidence intervals for 
#' standardized rates based on the approximate bootstrap method', 
#' *Statistics in Medicine*, **14**, 1875--1888.
#' \doi{doi:10.1002/sim.4780141704}. 
#' 
#' Fay MP & Feuer EJ (1997). 'Confidence intervals for directly 
#' standardized rates: a method based on the gamma distribution.
#' Statistics in Medicine*. **16**: 791--801.
#' \url{https://doi.org/10.1002/(SICI)1097-0258(19970415)16:7<791::AID-SIM500>3.0.CO;2-\%23}
#' 
#' Tiwari RC, Clegg LX, & Zou Z (2006). 'Efficient interval estimation 
#' for age-adjusted cancer rates.' 
#' *Statistical Methods in Medical Research* **15**: 547--569. 
#' \doi{doi:10.1177/0962280206070621}
#' 
#' Ng HKT, Filardo, G & Zheng G (2008). 'Confidence interval estimating 
#' procedures for standardized incidence rates.' 
#' *Computational Statistics and Data Analysis* **52** 3501--3516.
#' \doi{doi:10.1016/j.csda.2007.11.004}
#' @param x a vector of strata-specific counts. 
#' @param n a vector of strata-specific time bases for counts.
#' @param w a vector of strata-specific weights (or standard populations).
#' @param null.value a null hypothesis value of the directly 
#'  rate, if NULL no test is done. If not NULL, provide in rate per mult.
#' @param alternative type of alternative hypothesis.
#' @param conf.level confidence level for the returned 
#' confidence interval.
#' @param mult a factor to multiply the estimate and 
#' confidence intervals by, to give rates per mult.
#' @param method Method used to perform the test and construct the 
#' confidence interval. See details.
#' @param control list of arguments / type of modification used for
#' each method. See details and relevant `"xxxxControl"` documentation
#' @return a list with class `"htest"` containing the following 
#' components: 
#' \item{statistic}{number of strata or summands: 
#' \code{k = length(x)}}
#' \item{parameter}{mult}
#' \item{p.value}{p-value, set to \code{NA} if \code{null.value = NULL}}
#' \item{conf.int}{confidence interval on the true directly
#' standardized rate}
#' \item{estimate}{directly standardized rate}
#' \item{null.value}{null hypothesis value for the DSR}
#' \item{alternative}{alternative hypothesis type}
#' \item{method}{description of the method}
#' \item{data.name}{description of the data}
#' 
#' @importFrom stats qnorm qlnorm qbeta pbeta pnorm uniroot qlogis plogis
#' @importFrom exactci poisson.exact
#' @importFrom asht wspoissonTest
#' @importFrom loglognorm qloglognorm ploglognorm
#' @export
dsrTest <- function (x, n, w, null.value = NULL,
  alternative = c("two.sided", "less", "greater"),
  conf.level = 0.95, mult = 1,
  method = c("gamma", "asymptotic", "dobson", "beta", "bootstrap"),
  control = list()){
  # Functions for use later
  ciBound <- function(alternative, ci){
    # return the appropriate CI from c(alpha, 1-alpha)
    switch(alternative,
        less = c(0, ci[2]), greater = c(ci[1], Inf), two.sided = ci)
  }
  pz <- function(alternative, null.value, mean, sd){
  # return p.value from zscore (or NA where null.value is NULL)
      p.value <- NA
    if (!is.null(null.value)){
      zstat <- (null.value - mean) / sd
      p.value <- switch(alternative,
        less = stats::pnorm(zstat, lower.tail = FALSE),
        greater = stats::pnorm(zstat, lower.tail = TRUE),
        two.sided = 2 * (stats::pnorm(-abs(zstat))))
    }
  p.value
  }
  scaleNull <- function(null.value, mult, nullv = NULL, fun = identity){
  # scale null.value if not null
      r <- if (is.null(null.value)) nullv else fun(null.value / mult)
      r
  }
  methodName <- function(using, ...){
  # create a method name  (save a bit of typing)
    paste("Directly standardized rate: ", using, ..., sep = " ")
  }
  # - Argument checking
  alternative <- match.arg(alternative)
  method <- match.arg(method)
  if (conf.level < 0.5)
    stop("conf.level must be greater than 0.5")
  alpha <- 1 - conf.level
  if (alternative == "two.sided")
    alpha <- alpha / 2
  if (length(x) != length(w) || length(x) != length(n))
    stop("'x', 'n' and 'w' must be of the same length")
  if (anyNA(x))
    stop("missing values in 'x'")
  if (anyNA(n))
    stop("missing values in 'n'")
  if (anyNA(w))
    stop("missing values in 'w'")
  # controls
  control <- switch(method,
    dobson = do.call(dobsonControl, control),
    asymptotic = do.call(asymptoticControl, control),
    gamma = do.call(gammaControl, control),
    beta = do.call(betaControl, control),
    control)
  # estimated values
  # rescale weights to include timebases
  W <- (w / sum(w)) / n
  # dsr
  y <- sum(W * x)
  # variance of the dsr
  v <- sum(x * W ^ 2)
  # sum of x
  X <- sum(x)
  # CI and p values by method
  # -- Dobson - uses exactci::poisson.exact
  if (method == "dobson"){
    # Inverse (scale and shift) the null.value
    # exactci::poisson.exact requires non-NULL r
    nv <-
      if (is.null(null.value))
        1
    else
      X + ( (null.value / mult) - y) * sqrt(X / v)
    htest <- do.call(exactci::poisson.exact,
        c(list(x = X, r = nv, alternative = alternative,
          conf.level = conf.level), control))
    CINT <- ciBound(alternative,
                    y + sqrt(v / X ) * (htest[["conf.int"]] - X))
    p.value <- if (is.null(null.value)) NA else htest[["p.value"]]
    dname <- htest[["data.name"]]
    method <- methodName(
      "Dobson's method for Weighted Sum of Poissons with",
      htest[["method"]])
  }
  # Asymptotic
   if (method == "asymptotic"){
     control <- do.call(asymptoticControl, control)
     if (control[["trans"]] == "none"){
       CINT <- ciBound(alternative,
                stats::qnorm(c(alpha, 1 - alpha), y, sqrt(v)))
       p.value <- pz(alternative,
                     scaleNull(null.value, mult), y, sqrt(v))
       method <- methodName("Asymptotic method for Weighted Sum of Poissons",
                            "normal approximation of MLE")
     }
     if (control[["trans"]] == "log"){
       # variance of ln[y]
       vstar <- v / (y ^ 2)
       CINT <- ciBound(alternative,
        stats::qlnorm(c(alpha, 1 - alpha), log(y), sqrt(vstar)))
       zlstat <- scaleNull(null.value, mult, fun = log)
       p.value <- pz(alternative, zlstat, log(y), sqrt(vstar))
       method <- methodName("Asymptotic method for Weighted Sum of Poissons",
        "normal approximation of the log-transformed MLE")
     }
     if (control[["trans"]] == "loglog"){
       # variance of ln[ ln[-y]]
       vstar <- v / ( (y * log(y)) ^ 2)
       llog <- function(x) log(-log(x))
       CINT <- ciBound(alternative,
         loglognorm::qloglognorm(
           c(alpha, 1 - alpha), llog(y), sqrt(vstar)))
       nv <- scaleNull(null.value, mult)
       p.value <-
         if (is.null(null.value))
           NA
       else {
         pAL <- loglognorm::ploglognorm(nv, llog(y), sqrt(vstar))
         switch(alternative,
                less = 1 - pAL, greater = pAL,
                two.sided = min(1, 2 * pAL, 2 * pAG))
       }
       method <- methodName(
         "Asymptotic method for Weighted Sum of Poissons",
         "normal approximation of the log-log-transformed MLE")
     }
     if (control[["trans"]] == "logit"){
       # variance of logit[y]]
       vstar <- v / ( (y * (1 - y)) ^ 2)
       ql <- stats::qlogis
       CINT <- ciBound(alternative,
                 stats::plogis(stats::qnorm(
                 c(alpha, 1 - alpha), ql(y), sqrt(vstar))))
       zlstat <- scaleNull(null.value, mult, fun = ql)
       p.value <- pz(alternative, zlstat, ql(y), sqrt(vstar))
       method <- methodName(
         "Asymptotic method for Weighted Sum of Poissons",
         "normal approximation of the logit-transformed MLE")
     }
  }
  # gamma
  # uses asht::wsPoissonTest
  if (method == "gamma"){
    nv <- scaleNull(null.value, mult)
    htest <- do.call(asht::wspoissonTest, c(
      list(x = x, w = W, nullValue = nv, alternative = alternative,
           conf.level = conf.level), control))
    CINT <- htest[["conf.int"]]
    pv <- htest[["p.value"]]
    # approximately deal with asht issue that should be fixed soon
    # (only for midp = TRUE)
    p.value <- pv
    method <- methodName(htest[["method"]])
  }
  # beta
  if (method == "beta"){
    # different wmtype
    if (control[["wmtype"]] == "none"){
      yb <- y
      vb <- v
      modifierText <- ""
    }
    if (control[["wmtype"]] == "tcz"){
      yb <- y + mean(W)
      vb <- v + mean(W ^ 2)
      modifierText <- "[with Tiwari-Clegg-Zou modification]"
    }
    if (control[["wmtype"]] == "mean"){
      .wm <- mean(W)
      yb <- y + .wm
      vb <- v + .wm ^ 2
      modifierText <- "[with wm=mean(w) modification]"
    }
    if (control[["wmtype"]] == "minmaxavg"){
      .wm <- mean(c(min(W), max(W)))
      yb <- y + .wm
      vb <- v +  .wm ^ 2
      modifierText <- "[with wm=mean(min(w),max(w)) modification]"
    }
    if (control[["wmtype"]] == "max"){
      .wm <- max(W)
      yb <- y + .wm
      vb <- v + .wm ^ 2
      modifierText <- "[with wm=max(w) modification]"
    }
    a <- yb * (yb * (1 - yb) / vb - 1)
    b <- (1 - yb) * (yb * (1 - yb) / vb - 1)
    nv <- scaleNull(null.value, mult)
    CINT <- ciBound(alternative,
                    stats::qbeta(c(alpha, 1 - alpha), a, b))
    p.value <-
      if (is.null(null.value))
        NA
     else {
       pAL <- stats::pbeta(nv, a, b, lower.tail = FALSE)
       pAG <- stats::pbeta(nv, a, b, lower.tail = TRUE)
       switch(alternative,
         less = pAL, greater = pAG,
         two.sided = min(1, 2 * pAL, 2 * pAG))
     }
    method <- methodName("Beta method for Weighted Sum of Poissons",
                         modifierText)
  }
  # - Approximate bootstrap
  if (method == "bootstrap"){
    z0 <- a <- sum(x * (W ^ 3)) / (6 * v ^ (3 / 2))
    pFunction <- function(p){
      num <- z0 + stats::qnorm(p)
      y  + sqrt(v) * num / ( (1 - a * num) ^ 2)
    }
    CINT <- ciBound(alternative, pFunction(c(alpha, 1 - alpha)))
    if (is.null(null.value)){
      p.value <- NA
    } else {
      nv <- scaleNull(null.value, mult)
      f.AL <- function(p) pFunction(1 - p) - nv
      f.AG <- function(p)  (pFunction(p) - nv)
      pAL <- stats::uniroot(f.AL, interval = c(1e-09, 1 - 1e-09))$root
      pAG <- stats::uniroot(f.AG, interval = c(1e-09, 1 - 1e-09))$root
      p.value <- switch(alternative, less = pAL, greater = pAG,
        two.sided = min(1, 2 * pAG, 2 * pAL))
    }
    method <- methodName(
      "Approximate bootstrap method for Weighted Sum of Poissons")
  }
  # Set up htest attributes
  attr(CINT, "conf.level") <- conf.level
  names(y) <- "Directly standardarized rate"
  if (mult != 1)
    names(y) <- paste0("Directly standardized rate per ", mult)
  k <- length(x)
  names(k) <- "number of summands"
  parms <- mult
  names(parms)  <- "Multiplier"
  if (!is.null(null.value))
    names(null.value) <- "Directly standardized rate"
  xname <- paste0("x = ", deparse(substitute(x)))
  timebasename <- paste0("time bases: n = ", deparse(substitute(n)))
  wname <- paste0("weights: w = ", deparse(substitute(w)))
  dname <- paste(xname, timebasename, wname, sep = ", ")
  # returned object
  structure(
    list(
      statistic = k,
      parameter = parms,
      p.value = p.value,
      conf.int = CINT * mult,
      estimate = y * mult,
      null.value = null.value,
      alternative = alternative,
      method = method,
      data.name = dname
    ),
    class = "htest")
}
